File: MATEMÁTICA/2149/main.dart
import 'dart:io';

int phillBonati(int n, List<int> memo) {
  if (memo[n] == 0) {
    if (n < 2)
      memo[n] = n;
    else if (n.isOdd)
      memo[n] = phillBonati(n - 1, memo) * phillBonati(n - 2, memo);
    else
      memo[n] = phillBonati(n - 1, memo) + phillBonati(n - 2, memo);
  }
  return memo[n];
}

void main() {
  List<int> memo = List.filled(18, 0);
  int n;
  while ((n = int.parse(stdin.readLineSync() ?? '0')) > 0) {
    print(phillBonati(n - 1, memo));
  }
}

File: MATEMÁTICA/2149/main.rb
def phill_bonati(n, memo)
  if memo[n].zero?
    memo[n] = if n < 2
      n
    elsif n.odd?
      phill_bonati(n - 1, memo) * phill_bonati(n - 2, memo)
    else
      phill_bonati(n - 1, memo) + phill_bonati(n - 2, memo)
    end
  end
  memo[n]
end

memo = [0] * 17
$stdin.map(&:to_i).each do |n|
  puts phill_bonati(n - 1, memo)
end
